### PR TITLE
5 - Clone module

### PR DESCRIPTION
**Pull request by TAdev0**

### CODE REVIEW

File: corelib/src/clone.cairo [hunk 1]
```diff
@@ -1,9 +1,42 @@
...
+//! values, which can be particularly useful when working with types that do not implement
+//! the `Copy` trait or when you need to ensure deep copying of complex structures.
+//!
+//! # Examples
+//!
+//! ```
+//! let arr = array![1, 2, 3];
+//! let cloned_arr = arr.clone();
+//! assert!(arr == cloned_arr);
+//! ```
```
Comment (enitrat) [inline corelib/src/clone.cairo:12]:
```suggestion
//! The `Clone` trait provides the ability to duplicate instances of types that cannot be
//! 'implicitly copied'.
//!
//! In Cairo, some simple types are "implicitly copyable" and when you assign them or pass them as
//! arguments, the receiver will get a copy, leaving the original value in place. These types do not
//! require allocation to copy, and are not at risk of accessing un-allocated memory, so the
//! compiler considers them cheap and safe to copy. For other types, copies must be made explicitly,
//! by convention implementing the [`Clone`] trait and calling the [`clone`] method.
//!
//! # Examples
//!
//! ```
//! let arr = array![1, 2, 3];
//! let cloned_arr = arr.clone();
//! assert!(arr == cloned_arr);
//! ```
//!
//! To easily implement the Clone trait, you can also use
//! `#[derive(Clone)]`:
//! ```
//! #[derive(Clone, Drop)]
//! struct Sheep {
//!    name: ByteArray,
//!    age: u8,
//! }
//!
//! fn main() {
//!    let dolly = Sheep {
//!        name: "Dolly",
//!        age: 6,
//!    };
//!
//!    let cloned_sheep = dolly.clone();  // Famous cloned sheep!
//!}
//! ```
```

File: corelib/src/clone.cairo [hunk 2]
```diff
@@ -1,9 +1,42 @@
...
+    /// ```
+    ///
+    /// Note that explicitly passing a snapshot to `clone` is not mandatory, as the compiler is able
+    /// to infer that a snapshot is passed.
+    ///
+    /// ```
+    /// let value: u8 = 0;
+    /// let clone = value.clone();
+    /// assert!(clone == value);
+    /// ```
```
Comment (enitrat) [inline corelib/src/clone.cairo:39]:
```suggestion
    /// let value: u8 = 0;
    /// let clone = value.clone();
    /// assert!(clone == value);
```

don't cast to a snapshot unless strictly necessary; it's the compiler's job and should be transparent to the developper.

File: corelib/src/clone.cairo [hunk 3]
```diff
@@ -16,6 +49,24 @@ impl TupleClone<
...
+    ///
+    /// Note that explicitly passing a snapshot to `clone` is not mandatory, as the compiler is able
+    /// to infer that a snapshot is passed.
+    ///
+    /// ```
+    /// let tuple: (u8, u32) = (0, 1);
+    /// let clone = tuple.clone();
+    /// assert!(clone == tuple);
+    /// ```
```
Comment (enitrat) [inline corelib/src/clone.cairo:69]:
```suggestion
    /// # Examples
    ///
    /// ```
    /// let tuple: (u8, u32) = (0, 1);
    /// let clone = tuple.clone();
    /// assert!(clone == tuple);
    /// ```
```

**Review by enitrat — COMMENTED** [review]
Would globally benefit from more info on the clone system and why it's required, how it differs from Copy etc. the Suggestions help go into that direction, is fine for now

**Comment by TAdev0** [comment]
@enitrat  i modified a bit, to just add comment to the function definition in the trait
also commented with `//` elements that should not be present in the doc as suggested Orizi in fmt module PR

**Pull request closed without merging**